src/main/java/fr/sncf/osrd/train/ActivateRoute.java

```diff
@@ -17,22 +17,30 @@ public static void reserveRoutes(
             Simulation sim,
             Train train
     ) throws SimulationError {
-        // TODO have a smarter way to reserve routes
         var lastState = train.getLastState();
-        // If we have requested the last route of the path, we do nothing.
-        if (lastState.requestedRouteIndex + 1 >= lastState.path.routePath.size())
-            return;
-        // If we have requested at least two routes in advance, we do nothing.
-        if (lastState.routeIndex + 1 < lastState.requestedRouteIndex)
-            return;
-        // If a route has been reserved in advance and this route contains at least two tvdSection, we do nothing
-        if (lastState.routeIndex + 1 == lastState.requestedRouteIndex
-                && train.schedule.plannedPath.routePath.get(lastState.requestedRouteIndex).tvdSectionsPaths.size() > 1)
-            return;
-        var nextRoute = lastState.path.routePath.get(lastState.requestedRouteIndex + 1);
-        var nextRouteState = sim.infraState.getRouteState(nextRoute.index);
-        sim.infraState.towerState.request(sim, nextRouteState, train);
-        reserveNextIfNeeded(sim, train, lastState.requestedRouteIndex);
+        var path = lastState.path;
+        for (var routeIndex = lastState.routeIndex + 1; routeIndex < path.routePath.size(); routeIndex++) {
+            var route = path.routePath.get(routeIndex);
+            var routeState = sim.infraState.getRouteState(route.index);
+            if (!route.isControlled) {
+                // passive route
+                if (routeState.status.equals(RouteStatus.OCCUPIED)) {
+                    // occupied, we can't look any further
+                    break;
+                }  // otherwise, not occupied, we can keep looking forward
+
+            } else {
+                // controlled route
+                final var towerState = sim.infraState.towerState;
+                if (!towerState.isCurrentRouteAllowedForTrain(route, train.schedule.trainID)) {
+                    // we need to request it
+                    sim.infraState.towerState.request(sim, routeState, train);
+                    // we stop looking any further
+                    break;
```

Comment:
Ideally the function `towerState.request` return whether the request is approved or denied (and put in a waiting list). With this information we can possibly continue the loop and ask to reserve further routes.

Note: This could be done in the PR #241